Data Integrity During On-Boarding - Best Practices

mrjones-plip opened the pull request.

mrjones-plip commented on content/en/apps/guides/data/on-boarding-data-integrity.md:
````diff
@@ -0,0 +1,133 @@
+    ```
+    cat uuids-2.txt | jq --raw-input --slurp '{keys: split("\n") }' | curljz -d@- "$COUCH_URL/medic/_all_docs?include_docs=true" | jq '{docs: [ .rows[] | select(.doc).doc ]}' > docs.json
+    ```
+7. Change the fields by deleting "_rev" field and also "_attachments" if uploading in a different instance.
````
@Marina-Kenf - thanks for your input to answer @binokaryg question on this in the draft of this document. How does this look?

```suggestion
7. Change the fields by deleting "_rev" field and also "_attachments" if uploading in a different instance. This can be carefully done by manually editing the JSON locally before uploading or by running `DELETE` calls against the CouchDB database.
```

Marina-Kenf replied:
@mrjones-plip, I'll rather remove the deletion part in CouchDB unless we want to destroy the original documents on the initial instance.

mrjones-plip replied:
Awesome - thanks @Marina-Kenf ! I've updated the text and will commit the change.  

@binokaryg - lemme know if this satisfies your original "how" question!

mrjones-plip commented:
@abbyad  - I struggled as to where to add this on the docs site - open to moving it if you think it ended up in the wrong spot!

abbyad commented on content/en/apps/guides/data/on-boarding-data-integrity.md:
```diff
@@ -0,0 +1,133 @@
+
+![App URL in settings](app.url.png)
+
+The URL may not be a good proxy for CHWs to know which instance they're on. Will a CHW looking at  "example.**dev**.medicmobile.org" be able to tell the difference from "example.**app**.medicmobile.org"? To aid  CHWs, consider using a [different icon for the launcher and login]({{< relref "core/guides/android/branding" >}}).
```
It would be nice to avoid having medicmobile.org in here, so perhaps `project.dev.example.com`

mrjones-plip replied:
agreed! fixed in 3d1f96b .

abbyad commented on content/en/apps/guides/data/on-boarding-data-integrity.md:
```diff
@@ -0,0 +1,133 @@
+
+The URL may not be a good proxy for CHWs to know which instance they're on. Will a CHW looking at  "example.**dev**.medicmobile.org" be able to tell the difference from "example.**app**.medicmobile.org"? To aid  CHWs, consider using a [different icon for the launcher and login]({{< relref "core/guides/android/branding" >}}).
+
+**NB:** A caveat to using the hamburger menu is that the "Logout" option is here. CHWs should be cautious not to click this. Alternatively, "Logout" can be removed [from this menu via `can_log_out_on_android`]({{< relref "apps/reference/app-settings/user-permissions" >}}).
```
Although the informal tone is appropriate for a guide, the expressions like `is here` and `click this` make it hard to understand. Also, it's better to [avoid latin word/phrases](https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#latin-phrases) like `NB` and even `caveat`, and use a [note box](https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#notes-and-tips).

```suggestion
{{% alert title="Note" %}}Careful when selecting **About** to not select **Logout**. If accidental log outs are a frequent problem the logout option can be removed [from the menu by setting `can_log_out_on_android` to false]({{< relref "apps/reference/app-settings/user-permissions" >}}).{{% /alert %}}
```

abbyad commented on content/en/apps/guides/data/on-boarding-data-integrity.md:
```diff
@@ -0,0 +1,133 @@
+
+## Introduction
+
+When on-boarding CHWs or refreshing their training, it is important to both enable them to learn how to use the CHT using a real mobile device with a real CHT Android application on the exact same configuration as production will use. This will give them the confidence to correctly and safely use the app to do their first household visits. This guide assumes that CHWs will have both a training an ...
```
I keep waiting for part two of that first sentence because of the "important to both...."

mrjones-plip replied:
that sentence is too long as it is!! removed the "both" in 7fe4410

abbyad commented on content/en/apps/guides/data/on-boarding-data-integrity.md:
```diff
@@ -0,0 +1,133 @@
+
+---
+
+{{% pageinfo %}}
+This best practices guide should not be followed verbatim.  Instead, it is important to review each suggestion and assess if your deployment can benefit from it.
+{{% /pageinfo %}}
```
This block may be better as a note, but we can revisit after the main structure/overview is addressed.

abbyad reviewed (changes requested):
Nice to see this coming together!

When reading the doc I feel as though I am reading a troubleshooting guide, but missing the instructions. I found myself asking _How do we recommend CHWs be trained?_, which is important context for this guide. Can that be included before discussing the ways it can go wrong and remediation, and make this a broader _"Best practices for onboarding users"_ guide? If it is a single page, then that would mean moving everything over to something like `guides/onboarding`, but could also be split into two guides so that the remediation stands along in `guides/data`. Either way, I think that having that intro will help with the overall structure and flow for this content.

I will review the text more closely on the next pass, but left some comments that could be addressed now. Also, as a general note there seems to be a tendency to use `onboarding`, rather than `on-boarding`, so would use `onboard` throughout. That would also be consistent with [our use of it in the docs site](https://docs.communityhealthtoolkit.org/apps/examples/learning-care/).

kennsippell commented:
Is this assuming that each app has two APKs: a production APK and a training APK? In practice, I've seen deployments using the [medic app](https://play.google.com/store/apps/details?id=org.medicmobile.webapp.mobile&hl=en_CA&gl=US) on playstore for the training app.  then uninstalling the medic app after the training is completed.

mrjones-plip commented:
>  When reading the doc I feel as though I am reading a troubleshooting guide, but missing the instructions. I found myself asking How do we recommend CHWs be trained?, which is important context for this guide. Can that be included before discussing the ways it can go wrong and remediation, and make this a broader "Best practices for onboarding users" guide? If it is a single page, then that would mean moving everything over to something like guides/onboarding, but could also be split into two guides so that the remediation stands along in guides/data. Either way, I think that having that intro will help with the overall structure and flow for this content.

I'm very open to blocking the release of this page based on missing, broader "Best practices for onboarding users" content! However, such a guide is not one I'd feel comfortable writing, as I've never been through an onboarding, let alone to the field. I would be very open to curating this content by working with teammates who have the expertise!  

Maybe we could just add single sentence (or two!) to get this PR through and open another ticket to extend it to be more broad later? I'll try my hand, but am still cool to nix my sentences for a more proper solution per above.

mrjones-plip commented:
> Is this assuming that each app has two APKs: a production APK and a training APK? In practice, i'm unsure which projects actually do that. I've seen deployments using the [medic app](https://play.google.com/store/apps/details?id=org.medicmobile.webapp.mobile&hl=en_CA&gl=US) on playstore for the training app. then uninstalling the medic app after the training is completed.

Right - the scenario where the training app has a different icon for the launcher and for the login, I was thinking it would be two different APKs and not using the [unbranded app](https://play.google.com/store/apps/details?id=org.medicmobile.webapp.mobile&hl=en_CA&gl=US) .  That said, I think using the unbranded app is fine alternative in a pinch!  The draw back would be that if you log out, you now need your username, password AND the URL for the dev instance.

do you think we should add some clarifying content around this?

mrjones-plip commented:
> as a general note there seems to be a tendency to use onboarding, rather than on-boarding, so would use onboard throughout. That would also be consistent with [our use of it in the docs site](https://docs.communityhealthtoolkit.org/apps/examples/learning-care/).

Thanks @abbyad! fixed in f7190cc

mrjones-plip commented:
> Maybe we could just add single sentence (or two!) to get this PR through and open another ticket to extend it to be more broad later? I'll try my hand, but am still cool to nix my sentences for a more proper solution per above.

@abbyad  - I gave it a try! Kinda feels like it's not enough...but maybe you think 32679fa is totally awesome!?

mrjones-plip commented:
> Ah - this just got edited, thanks for the extra info! I'll report back with a commit hash so you can see what I added. Thanks for the input!

@kennsippell  - I've added more info in 7372828 - please let me know what you think!

mrjones-plip commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+- **App name**: Provide a noticeably different name to the training app. Since app names are often cut short on Android devices, make the change at beginning of the text. For example, `[TRAINING] CHW App` would be better than `CHW App [TRAINING VERSION]`, since the latter may only display as `CHW App...`.
+- **Border & Message**: Consider adding a distinctive border and message when using the training app. This might be done in a more automated way in future versions of CHT Android, and in the meantime can be [modelled after this example]().
+
+![CHT with border](border.png)
```
minor tweak: we should prepend three spaces so this renders as part of the above bullet point:
```suggestion
   ![CHT with border](border.png)
```

abbyad replied:
Good eye!

mrjones-plip commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+### Setting up a training app
+
+A separate Android App can be created for training, which would point to a CHT instance dedicated to training. The training instance should have the same configuration as the production instance, and have users created for training. To differentiate the Android app used for training from the production one, create a duplicate of your brand, and modify the following aspects:
```
We should deep link to the technical how to and I think use "flavor" over "brand" to stick with existing vernacular:

```suggestion
A separate Android App can be created for training, which would point to a CHT instance dedicated to training. The training instance should have the same configuration as the production instance, and have users created for training. To differentiate the Android app used for training from the production one, create a duplicate of [your flavor]({{< relref "core/guides/android/branding" >}}) and modify the following aspects
```

abbyad commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+When onboarding new users, having a dedicated CHT app and instance for training can be helpful; it allows new users to do training exercises with mock data to get familiar with the app, while not having the data from their training interfering with the future use of their CHT app. Different approaches are possible, such as only entering real patient data during training, or manually deleting all  ...
+
+{{% alert title="Note" %}}The suggestions in this guide should be assessed and adapted as needed to benefit a deployment. It is important that users don't accidentally use the wrong app. The [troubleshooting guide]() can help to monitor and remediate training data being in the production instance, or the opposite.{{% /alert %}}
```
```suggestion
{{% alert title="Note" %}}The suggestions in this guide should be assessed and adapted as needed to benefit a deployment. It is important that users don't accidentally use the wrong app. The [troubleshooting guide]({{< relref "apps/guides/data/fixing-data-in-wrong-location" >}}) can help to monitor and remediate training data being in the production instance, or the opposite.{{% /alert %}}
```

abbyad commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+Changing passwords for the training users in an attempt to lock them out is not recommended. In some circumstances a user would be able to continue to use the training app for production use and not have the data sync back to the server.
+
+It is preferable to remove the training app from devices, and [monitor the training instance for unexpected activity]() that can be brought over to the production instance if needed.
```
```suggestion
It is preferable to remove the training app from devices, and [monitor the training instance for unexpected activity](< relref "apps/guides/data/fixing-data-in-wrong-location" >}}) that can be brought over to the production instance if needed.
```

abbyad commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+- **CHT Instance**: Set the `app_host` string to be the URL of your training instance, otherwise it will in effect be the same as using the production app. 
+- **Launcher icons**: Consider using completely different icons, or at least change the color of the launcher icons.
+- **App name**: Provide a noticeably different name to the training app. Since app names are often cut short on Android devices, make the change at beginning of the text. For example, `[TRAINING] CHW App` would be better than `CHW App [TRAINING VERSION]`, since the latter may only display as `CHW App...`.
+- **Border & Message**: Consider adding a distinctive border and message when using the training app. This might be done in a more automated way in future versions of CHT Android, and in the meantime can be [modelled after this example]().
```
```suggestion
- **Border & Message**: Consider adding a distinctive border and message when using the training app. This might be done in a more automated way in future versions of CHT Android, and in the meantime can be [modeled after this example](https://github.com/medic/cht-android/commit/8d077ed08dc3889ef1f4e3bad7231931bca55d87).
```

abbyad reviewed (approved):
This came together really nicely, thanks @mrjones-plip for pushing this through!

abbyad commented:
Just noticed that the paths need updating, along with header alignment. We should update those before merging.

abbyad commented on content/en/apps/guides/onboarding/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+{{% alert title="Note" %}}The suggestions in this guide should be assessed and adapted as needed to benefit a deployment. It is important that users don't accidentally use the wrong app. The [troubleshooting guide]({{< relref "apps/guides/data/fixing-data-in-wrong-location" >}}) can help to monitor and remediate training data being in the production instance, or the opposite.{{% /alert %}}
+
+### Setting up a training app
```
```suggestion
## Setting up a training app
```

abbyad commented on content/en/apps/guides/data/fixing-data-in-wrong-location.md:
````diff
@@ -0,0 +1,104 @@
+    }
+   ```
+
+    ![Creating a view in Fauxton](create.view.png)
````
Path is broken. File should be moved accordingly.

mrjones-plip commented:
@kennsippell  - Imma push this out, but feel free to open a ticket for me to fix anything upon my return!

Thanks @abbyad  for the great work session today, much obliged.
